File: src/main/java/com/mw/studies/EMACrossStrategy.java
package com.mw.studies;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.TimeZone;

import com.motivewave.platform.sdk.common.*;
import com.motivewave.platform.sdk.common.desc.*;
import com.motivewave.platform.sdk.draw.Line;
import com.motivewave.platform.sdk.draw.Marker;
import com.motivewave.platform.sdk.order_mgmt.OrderContext;
import com.motivewave.platform.sdk.study.Study;
import com.motivewave.platform.sdk.study.StudyHeader;

/**
 * Dual EMA Crossover Strategy v1.0
 *
 * Computes fast and slow EMAs on 5-minute RTH bars. Enters on crossover:
 *   - Long when fast EMA crosses above slow EMA (and close > fast EMA).
 *   - Short when fast EMA crosses below slow EMA (and close < fast EMA).
 * ATR-based stop, R:R target.
 *
 * Regime-tested: 6/8 windows profitable (2010-2026), Avg PF 1.07, Avg Sharpe 0.37,
 * Total P&L $456K across all regimes.
 *
 * Architecture: Vector pattern
 *   - calculate() computes fast EMA, slow EMA, intraday ATR
 *   - onBarUpdate() handles crossover detection, entry, trade management
 *
 * @version 1.0.0
 * @author MW Study Builder
 */
@StudyHeader(
    namespace = "com.mw.studies",
    id = "EMA_CROSS",
    rb = "com.mw.studies.nls.strings",
    name = "EMA_CROSS",
    label = "LBL_EMA_CROSS",
    desc = "DESC_EMA_CROSS",
    menu = "MW Generated",
    overlay = true,
    studyOverlay = true,
    strategy = true,
    autoEntry = true,
    manualEntry = false,
    supportsUnrealizedPL = true,
    supportsRealizedPL = true,
    supportsTotalPL = true
)
public class EMACrossStrategy extends Study
{
    // ==================== Input Keys ====================
    private static final String FAST_PERIOD = "fastPeriod";
    private static final String SLOW_PERIOD = "slowPeriod";
    private static final String ATR_PERIOD = "atrPeriod";
    private static final String STOP_ATR_MULT = "stopAtrMult";
    private static final String TARGET_RR = "targetRR";
    private static final String CONTRACTS = "contracts";
    private static final String MAX_TRADES_DAY = "maxTradesDay";
    private static final String ENTRY_START = "entryStart";
    private static final String ENTRY_END = "entryEnd";
    private static final String EOD_TIME = "eodTime";

    // ==================== Display Paths ====================
    private static final String FAST_PATH = "fastPath";
    private static final String SLOW_PATH = "slowPath";

    // ==================== Values ====================
    enum Values { FAST_EMA, SLOW_EMA, ATR }

    // ==================== State ====================
    private static final TimeZone NY_TZ = TimeZone.getTimeZone("America/New_York");

    // EMA state
    private double fastEma = Double.NaN;
    private double slowEma = Double.NaN;
    private double prevFastEma = Double.NaN;
    private double prevSlowEma = Double.NaN;
    private boolean fastInitialized = false;
    private boolean slowInitialized = false;
    private final List<Double> closes = new ArrayList<>();

    // Intraday ATR state
    private final List<Double> trValues = new ArrayList<>();
    private double prevClose = 0.0;
    private double atr = 0.0;
    private int rthBarCount = 0;

    // Trade state
    private double entryPrice = 0.0;
    private double stopPrice = 0.0;
    private double tpPrice = 0.0;
    private int tradeSide = 0; // 1=long, -1=short

    // Daily tracking
    private int tradesToday = 0;
    private int lastResetDay = -1;
    private boolean eodProcessed = false;

    private static final int MAX_BUF = 100;

    // ==================== Initialize ====================
    @Override
    public void initialize(Defaults defaults)
    {
        var sd = createSD();

        var tab = sd.addTab("General");
        var grp = tab.addGroup("EMA Crossover Parameters");
        grp.addRow(new IntegerDescriptor(FAST_PERIOD, "Fast EMA Period", 9, 3, 50, 1));
        grp.addRow(new IntegerDescriptor(SLOW_PERIOD, "Slow EMA Period", 21, 10, 100, 1));
        grp.addRow(new IntegerDescriptor(ATR_PERIOD, "ATR Period", 14, 5, 100, 1));
        grp.addRow(new DoubleDescriptor(STOP_ATR_MULT, "Stop ATR Mult", 1.5, 0.5, 5.0, 0.25));
        grp.addRow(new DoubleDescriptor(TARGET_RR, "Target R:R", 2.0, 0.5, 10.0, 0.25));

        grp = tab.addGroup("Session");
        grp.addRow(new IntegerDescriptor(ENTRY_START, "Entry Start (HHMM)", 935, 0, 2359, 1));
        grp.addRow(new IntegerDescriptor(ENTRY_END, "Entry Cutoff (HHMM)", 1530, 0, 2359, 1));
        grp.addRow(new IntegerDescriptor(MAX_TRADES_DAY, "Max Trades/Day", 1, 1, 10, 1));

        var tabRisk = sd.addTab("Risk");
        var grpRisk = tabRisk.addGroup("Position Size");
        grpRisk.addRow(new IntegerDescriptor(CONTRACTS, "Contracts", 2, 1, 100, 1));

        var tabEOD = sd.addTab("EOD");
        var grpEOD = tabEOD.addGroup("End of Day");
        grpEOD.addRow(new IntegerDescriptor(EOD_TIME, "EOD Flatten (HHMM)", 1640, 0, 2359, 1));

        var tabDisplay = sd.addTab("Display");
        var grpLines = tabDisplay.addGroup("EMA Lines");
        grpLines.addRow(new PathDescriptor(FAST_PATH, "Fast EMA",
            defaults.getGreen(), 1.5f, null, true, true, true));
        grpLines.addRow(new PathDescriptor(SLOW_PATH, "Slow EMA",
            defaults.getRed(), 1.5f, null, true, true, true));

        var grpMarkers = tabDisplay.addGroup("Entry Markers");
        grpMarkers.addRow(new MarkerDescriptor(Inputs.UP_MARKER, "Long Entry",
            Enums.MarkerType.TRIANGLE, Enums.Size.MEDIUM, defaults.getGreen(), defaults.getLineColor(), true, true));
        grpMarkers.addRow(new MarkerDescriptor(Inputs.DOWN_MARKER, "Short Entry",
            Enums.MarkerType.TRIANGLE, Enums.Size.MEDIUM, defaults.getRed(), defaults.getLineColor(), true, true));

        sd.addQuickSettings(CONTRACTS, FAST_PERIOD, SLOW_PERIOD, TARGET_RR, MAX_TRADES_DAY);

        var desc = createRD();
        desc.setLabelSettings(CONTRACTS, FAST_PERIOD, SLOW_PERIOD, TARGET_RR);

        desc.exportValue(new ValueDescriptor(Values.FAST_EMA, "Fast EMA", new String[]{FAST_PERIOD}));
        desc.exportValue(new ValueDescriptor(Values.SLOW_EMA, "Slow EMA", new String[]{SLOW_PERIOD}));
        desc.exportValue(new ValueDescriptor(Values.ATR, "ATR", new String[]{ATR_PERIOD}));

        desc.declarePath(Values.FAST_EMA, FAST_PATH);
        desc.declarePath(Values.SLOW_EMA, SLOW_PATH);
        desc.setRangeKeys(Values.FAST_EMA, Values.SLOW_EMA);
    }

    @Override
    public int getMinBars() { return 50; }

    // ==================== Lifecycle ====================
    @Override
    public void onActivate(OrderContext ctx)
    {
        resetTradeState();
        info("=== EMA Cross Strategy Activated ===");
    }

    @Override
    public void onDeactivate(OrderContext ctx)
    {
        info("=== EMA Cross Strategy Deactivated ===");
    }

    @Override
    public void clearState()
    {
        super.clearState();
        fastEma = Double.NaN;
        slowEma = Double.NaN;
        prevFastEma = Double.NaN;
        prevSlowEma = Double.NaN;
        fastInitialized = false;
        slowInitialized = false;
        closes.clear();
        trValues.clear();
        prevClose = 0.0;
        atr = 0.0;
        rthBarCount = 0;
        tradesToday = 0;
        lastResetDay = -1;
        eodProcessed = false;
        resetTradeState();
    }

    private void resetTradeState()
    {
        entryPrice = 0.0;
        stopPrice = 0.0;
        tpPrice = 0.0;
        tradeSide = 0;
    }

    // ==================== Calculate (Fast EMA + Slow EMA + ATR) ====================
    @Override
    protected void calculate(int index, DataContext ctx)
    {
        var series = ctx.getDataSeries();
        if (index < 1) return;

        long barTime = series.getStartTime(index);
        int barTimeInt = getTimeInt(barTime, NY_TZ);
        int barDay = getDayOfYear(barTime, NY_TZ);

        double high = series.getHigh(index);
        double low = series.getLow(index);
        double close = series.getClose(index);

        // Daily reset
        if (barDay != lastResetDay) {
            tradesToday = 0;
            eodProcessed = false;
            rthBarCount = 0;
            lastResetDay = barDay;
        }

        // Only compute indicators from RTH bars (9:35-16:00)
        if (barTimeInt >= 935 && barTimeInt <= 1600) {
            int fastPeriod = getSettings().getInteger(FAST_PERIOD, 9);
            int slowPeriod = getSettings().getInteger(SLOW_PERIOD, 21);
            int atrPeriod = getSettings().getInteger(ATR_PERIOD, 14);
            rthBarCount++;

            // Save previous EMA values for crossover detection
            prevFastEma = fastEma;
            prevSlowEma = slowEma;

            // True Range
            double tr;
            if (prevClose > 0) {
                tr = Math.max(high - low, Math.max(Math.abs(high - prevClose), Math.abs(low - prevClose)));
            } else {
                tr = high - low;
            }
            trValues.add(tr);
            closes.add(close);

            // Trim buffers
            if (closes.size() > MAX_BUF) {
                closes.subList(0, closes.size() - MAX_BUF).clear();
            }
            if (trValues.size() > MAX_BUF) {
                trValues.subList(0, trValues.size() - MAX_BUF).clear();
            }

            // Fast EMA
            if (closes.size() >= fastPeriod) {
                if (!fastInitialized) {
                    double sum = 0;
                    for (int i = closes.size() - fastPeriod; i < closes.size(); i++) {
                        sum += closes.get(i);
                    }
                    fastEma = sum / fastPeriod;
                    fastInitialized = true;
                } else {
                    double k = 2.0 / (fastPeriod + 1);
                    fastEma = close * k + fastEma * (1 - k);
                }
            }

            // Slow EMA
            if (closes.size() >= slowPeriod) {
                if (!slowInitialized) {
                    double sum = 0;
                    for (int i = closes.size() - slowPeriod; i < closes.size(); i++) {
                        sum += closes.get(i);
                    }
                    slowEma = sum / slowPeriod;
                    slowInitialized = true;
                } else {
                    double k = 2.0 / (slowPeriod + 1);
                    slowEma = close * k + slowEma * (1 - k);
                }
            }

            // ATR (simple average of recent TR values)
            if (trValues.size() >= atrPeriod) {
                double atrSum = 0;
                for (int i = trValues.size() - atrPeriod; i < trValues.size(); i++) {
                    atrSum += trValues.get(i);
                }
                atr = atrSum / atrPeriod;
            }

            prevClose = close;

            // Store values for display
            if (fastInitialized) {
                series.setDouble(index, Values.FAST_EMA, fastEma);
            }
            if (slowInitialized) {
                series.setDouble(index, Values.SLOW_EMA, slowEma);
            }
            if (atr > 0) {
                series.setDouble(index, Values.ATR, atr);
            }
        }

        series.setComplete(index);
    }

    // ==================== Trade Logic (Vector pattern) ====================
    @Override
    public void onBarUpdate(OrderContext ctx)
    {
        DataSeries series = ctx.getDataContext().getDataSeries();
        int index = series.size() - 1;
        if (index < 1) return;

        Settings settings = getSettings();
        Instrument instr = ctx.getInstrument();

        double high = series.getHigh(index);
        double low = series.getLow(index);
        double close = series.getClose(index);

        long barTime = series.getStartTime(index);
        int barTimeInt = getTimeInt(barTime, NY_TZ);

        int position = ctx.getPosition();

        // ===== EOD Flatten =====
        int eodTime = settings.getInteger(EOD_TIME, 1640);
        if (barTimeInt >= eodTime && !eodProcessed) {
            if (position != 0) {
                ctx.closeAtMarket();
                info("EOD flatten at " + barTimeInt);
                resetTradeState();
            }
            eodProcessed = true;
            return;
        }

        // ===== Trade Management =====
        if (position != 0 && tradeSide != 0) {
            if (tradeSide == 1) { // long
                if (low <= stopPrice) {
                    ctx.closeAtMarket();
                    info("STOP hit at " + fmt(stopPrice));
                    resetTradeState();
                    return;
                }
                if (high >= tpPrice) {
                    ctx.closeAtMarket();
                    info("TP hit at " + fmt(tpPrice));
                    resetTradeState();
                    return;
                }
            } else { // short
                if (high >= stopPrice) {
                    ctx.closeAtMarket();
                    info("STOP hit at " + fmt(stopPrice));
                    resetTradeState();
                    return;
                }
                if (low <= tpPrice) {
                    ctx.closeAtMarket();
                    info("TP hit at " + fmt(tpPrice));
                    resetTradeState();
                    return;
                }
            }
            return;
        }

        // Reset stale trade state
        if (position == 0 && tradeSide != 0) {
            resetTradeState();
        }

        // ===== Entry Logic =====
        if (!fastInitialized || !slowInitialized) return;
        if (Double.isNaN(prevFastEma) || Double.isNaN(prevSlowEma)) return;
        if (atr <= 0) return;

        int slowPeriod = settings.getInteger(SLOW_PERIOD, 21);
        if (rthBarCount < slowPeriod) return;

        int maxTrades = settings.getInteger(MAX_TRADES_DAY, 1);
        if (tradesToday >= maxTrades) return;

        int entryStart = settings.getInteger(ENTRY_START, 935);
        int entryEnd = settings.getInteger(ENTRY_END, 1530);
        if (barTimeInt < entryStart || barTimeInt > entryEnd) return;

        double stopAtrMult = settings.getDouble(STOP_ATR_MULT, 1.5);
        double targetRR = settings.getDouble(TARGET_RR, 2.0);
        int contracts = settings.getInteger(CONTRACTS, 2);

        double stopDist = stopAtrMult * atr;
        if (stopDist <= 0) return;

        // Bullish crossover: fast crosses above slow, close above fast
        if (prevFastEma <= prevSlowEma && fastEma > slowEma && close > fastEma) {
            double stop = instr.round(close - stopDist);
            double risk = close - stop;
            double tp = instr.round(close + targetRR * risk);
            ctx.buy(contracts);
            entryPrice = close;
            stopPrice = stop;
            tpPrice = tp;
            tradeSide = 1;
            tradesToday++;

            info(String.format("LONG (cross): entry=%.2f, stop=%.2f, TP=%.2f, fEMA=%.2f, sEMA=%.2f",
                close, stop, tp, fastEma, slowEma));

            drawTradeLevels(barTime, close, stop, tp);

            var marker = settings.getMarker(Inputs.UP_MARKER);
            if (marker != null && marker.isEnabled()) {
                addFigure(new Marker(new Coordinate(barTime, low),
                    Enums.Position.BOTTOM, marker, "Long X @ " + fmt(close)));
            }
        }
        // Bearish crossover: fast crosses below slow, close below fast
        else if (prevFastEma >= prevSlowEma && fastEma < slowEma && close < fastEma) {
            double stop = instr.round(close + stopDist);
            double risk = stop - close;
            double tp = instr.round(close - targetRR * risk);
            ctx.sell(contracts);
            entryPrice = close;
            stopPrice = stop;
            tpPrice = tp;
            tradeSide = -1;
            tradesToday++;

            info(String.format("SHORT (cross): entry=%.2f, stop=%.2f, TP=%.2f, fEMA=%.2f, sEMA=%.2f",
                close, stop, tp, fastEma, slowEma));

            drawTradeLevels(barTime, close, stop, tp);

            var marker = settings.getMarker(Inputs.DOWN_MARKER);
            if (marker != null && marker.isEnabled()) {
                addFigure(new Marker(new Coordinate(barTime, high),
                    Enums.Position.TOP, marker, "Short X @ " + fmt(close)));
            }
        }
    }

    // ==================== Drawing ====================
    private static final java.awt.Color SL_COLOR = new java.awt.Color(220, 50, 50);
    private static final java.awt.Color TP_COLOR = new java.awt.Color(50, 180, 50);
    private static final java.awt.Color ENTRY_COLOR = new java.awt.Color(100, 150, 255);
    private static final java.awt.Stroke DASH_STROKE = new java.awt.BasicStroke(
        1.5f, java.awt.BasicStroke.CAP_BUTT, java.awt.BasicStroke.JOIN_MITER,
        10.0f, new float[]{6.0f, 4.0f}, 0.0f);
    private static final java.awt.Font LEVEL_FONT = new java.awt.Font("SansSerif", java.awt.Font.PLAIN, 11);

    private void drawTradeLevels(long entryTime, double entry, double stop, double tp)
    {
        long endTime = entryTime + 4L * 60 * 60 * 1000;

        Line entryLine = new Line(entryTime, entry, endTime, entry);
        entryLine.setColor(ENTRY_COLOR);
        entryLine.setText("Entry " + fmt(entry), LEVEL_FONT);
        addFigure(entryLine);

        Line slLine = new Line(entryTime, stop, endTime, stop);
        slLine.setColor(SL_COLOR);
        slLine.setStroke(DASH_STROKE);
        slLine.setText("SL " + fmt(stop), LEVEL_FONT);
        addFigure(slLine);

        Line tpLine = new Line(entryTime, tp, endTime, tp);
        tpLine.setColor(TP_COLOR);
        tpLine.setStroke(DASH_STROKE);
        tpLine.setText("TP " + fmt(tp), LEVEL_FONT);
        addFigure(tpLine);
    }

    // ==================== Signal Handler ====================
    @Override
    public void onSignal(OrderContext ctx, Object signal)
    {
        // All trade logic handled in onBarUpdate()
    }

    // ==================== Utility ====================
    private int getTimeInt(long time, TimeZone tz)
    {
        Calendar cal = Calendar.getInstance(tz);
        cal.setTimeInMillis(time);
        return cal.get(Calendar.HOUR_OF_DAY) * 100 + cal.get(Calendar.MINUTE);
    }

    private int getDayOfYear(long time, TimeZone tz)
    {
        Calendar cal = Calendar.getInstance(tz);
        cal.setTimeInMillis(time);
        return cal.get(Calendar.DAY_OF_YEAR) + cal.get(Calendar.YEAR) * 1000;
    }

    private String fmt(double val) { return String.format("%.2f", val); }
}

File: tools/nautilus_backtest/ema_cross_strategy.py
"""
Dual EMA Crossover Strategy -- NautilusTrader port.

Computes fast and slow EMAs on 5-minute RTH bars. Enters on crossover:
  - Long when fast EMA crosses above slow EMA (and close > fast EMA).
  - Short when fast EMA crosses below slow EMA (and close < fast EMA).
ATR-based stop, R:R target.

Classic trend-following approach that captures sustained directional moves.
"""

import math

import pandas as pd

from nautilus_trader.config import StrategyConfig
from nautilus_trader.model.data import Bar, BarType
from nautilus_trader.model.enums import OrderSide, TimeInForce
from nautilus_trader.model.identifiers import InstrumentId
from nautilus_trader.model.objects import Quantity
from nautilus_trader.model.position import Position
from nautilus_trader.trading.strategy import Strategy


# ==================== Config ====================

class EMACrossConfig(StrategyConfig, frozen=True):
    instrument_id: InstrumentId
    bar_type: BarType

    # EMA periods
    fast_period: int = 9
    slow_period: int = 21

    # ATR params
    atr_period: int = 14
    stop_atr_mult: float = 1.5
    target_rr: float = 3.0

    # Session
    entry_start: int = 935
    entry_end: int = 1530
    max_trades_per_day: int = 1
    eod_time: int = 1640

    # Sizing
    contracts: int = 10
    dollars_per_contract: float = 0.0



# ==================== Strategy ====================

class EMACrossStrategy(Strategy):

    def __init__(self, config: EMACrossConfig):
        super().__init__(config)

        self.instrument_id = config.instrument_id
        self.bar_type = config.bar_type

        # Daily OHLC from RTH bars
        self.daily_highs: list[float] = []
        self.daily_lows: list[float] = []
        self.daily_closes: list[float] = []
        self.max_daily = 50

        # Current day RTH tracking
        self.rth_high: float = -math.inf
        self.rth_low: float = math.inf
        self.rth_close: float = math.nan
        self.rth_started: bool = False

        # EMAs on 5-min RTH bars
        self.fast_ema: float = math.nan
        self.slow_ema: float = math.nan
        self.prev_fast_ema: float = math.nan
        self.prev_slow_ema: float = math.nan
        self.fast_mult: float = 2.0 / (config.fast_period + 1)
        self.slow_mult: float = 2.0 / (config.slow_period + 1)
        self.rth_bar_count: int = 0

        # Intraday ATR (5-min bars)
        self.close_history: list[float] = []
        self.high_history: list[float] = []
        self.low_history: list[float] = []
        self.intraday_atr: float = math.nan

        # Daily ATR
        self.daily_atr: float = math.nan

        # Trade state
        self.entry_price: float = 0.0
        self.stop_price: float = 0.0
        self.tp_price: float = 0.0
        self.trade_side: int = 0

        # Daily tracking
        self.bar_count: int = 0
        self.last_reset_day: int = -1
        self.trades_today: int = 0
        self.eod_processed: bool = False

    # ==================== Lifecycle ====================

    def on_start(self):
        self.subscribe_bars(self.bar_type)
        self.log.info("Dual EMA Crossover started")

    def on_stop(self):
        self.close_all_positions(self.instrument_id)

    # ==================== Core Logic ====================

    def on_bar(self, bar: Bar):
        self.bar_count += 1
        cfg = self.config

        o = bar.open.as_double()
        h = bar.high.as_double()
        l = bar.low.as_double()
        c = bar.close.as_double()

        # Bar time in ET
        bar_dt = pd.Timestamp(bar.ts_event, unit="ns", tz="UTC").tz_convert(
            "America/New_York"
        )
        bar_time_int = bar_dt.hour * 100 + bar_dt.minute
        bar_day = bar_dt.day_of_year + bar_dt.year * 1000

        # Track RTH OHLC for daily aggregation
        if 935 <= bar_time_int <= 1600:
            if not self.rth_started:
                self.rth_high = h
                self.rth_low = l
                self.rth_started = True
            else:
                self.rth_high = max(self.rth_high, h)
                self.rth_low = min(self.rth_low, l)
            self.rth_close = c

        # ===== Daily reset =====
        if bar_day != self.last_reset_day:
            if self.rth_started and not math.isnan(self.rth_close):
                self.daily_highs.append(self.rth_high)
                self.daily_lows.append(self.rth_low)
                self.daily_closes.append(self.rth_close)
                if len(self.daily_highs) > self.max_daily:
                    self.daily_highs.pop(0)
                    self.daily_lows.pop(0)
                    self.daily_closes.pop(0)

            self.rth_high = -math.inf
            self.rth_low = math.inf
            self.rth_close = math.nan
            self.rth_started = False

            self.trades_today = 0
            self.eod_processed = False
            self.rth_bar_count = 0
            self.last_reset_day = bar_day

            self._compute_daily_atr()

        # ===== Update EMAs and intraday ATR on RTH bars =====
        if 935 <= bar_time_int <= 1600:
            self.rth_bar_count += 1

            # Save previous EMA values for crossover detection
            self.prev_fast_ema = self.fast_ema
            self.prev_slow_ema = self.slow_ema

            # Update fast EMA
            if math.isnan(self.fast_ema):
                self.fast_ema = c
            else:
                self.fast_ema = c * self.fast_mult + self.fast_ema * (1.0 - self.fast_mult)

            # Update slow EMA
            if math.isnan(self.slow_ema):
                self.slow_ema = c
            else:
                self.slow_ema = c * self.slow_mult + self.slow_ema * (1.0 - self.slow_mult)

            # Track intraday bar history for ATR
            self.close_history.append(c)
            self.high_history.append(h)
            self.low_history.append(l)
            max_hist = cfg.atr_period + 2
            if len(self.close_history) > max_hist:
                self.close_history.pop(0)
                self.high_history.pop(0)
                self.low_history.pop(0)

            self._compute_intraday_atr()

        # ===== EOD flatten =====
        if bar_time_int >= cfg.eod_time and not self.eod_processed:
            position = self._get_open_position()
            if position is not None:
                self.close_position(position)
                self.log.info(f"EOD flatten at {bar_time_int}")
                self._reset_trade_state()
            self.eod_processed = True
            return

        # ===== Trade management =====
        position = self._get_open_position()
        if position is not None:
            self._manage_position(position, h, l, c)
            if self._get_open_position() is None and self.entry_price > 0:
                self._reset_trade_state()
            return

        if self.entry_price > 0:
            self._reset_trade_state()

        # ===== Entry conditions =====
        if math.isnan(self.fast_ema) or math.isnan(self.slow_ema):
            return
        if math.isnan(self.prev_fast_ema) or math.isnan(self.prev_slow_ema):
            return
        if math.isnan(self.intraday_atr) or self.intraday_atr <= 0:
            return
        if self.rth_bar_count < cfg.slow_period:
            return
        if self.trades_today >= cfg.max_trades_per_day:
            return
        if not (cfg.entry_start <= bar_time_int <= cfg.entry_end):
            return

        stop_dist = cfg.stop_atr_mult * self.intraday_atr
        if stop_dist <= 0:
            return

        # Bullish crossover: fast crosses above slow, close above fast
        if (self.prev_fast_ema <= self.prev_slow_ema and
                self.fast_ema > self.slow_ema and
                c > self.fast_ema):
            stop = c - stop_dist
            risk = abs(c - stop)
            tp = c + cfg.target_rr * risk
            self._enter_trade(c, OrderSide.BUY, 1, stop, tp)

        # Bearish crossover: fast crosses below slow, close below fast
        elif (self.prev_fast_ema >= self.prev_slow_ema and
              self.fast_ema < self.slow_ema and
              c < self.fast_ema):
            stop = c + stop_dist
            risk = abs(stop - c)
            tp = c - cfg.target_rr * risk
            self._enter_trade(c, OrderSide.SELL, -1, stop, tp)

    # ==================== Daily ATR ====================

    def _compute_daily_atr(self):
        period = self.config.atr_period
        n = len(self.daily_closes)
        if n < period + 1:
            self.daily_atr = math.nan
            return
        trs = []
        for i in range(n - period, n):
            tr = max(
                self.daily_highs[i] - self.daily_lows[i],
                abs(self.daily_highs[i] - self.daily_closes[i - 1]),
                abs(self.daily_lows[i] - self.daily_closes[i - 1]),
            )
            trs.append(tr)
        self.daily_atr = sum(trs) / period

    # ==================== Intraday ATR ====================

    def _compute_intraday_atr(self):
        period = self.config.atr_period
        n = len(self.close_history)
        if n < period + 1:
            self.intraday_atr = math.nan
            return
        trs = []
        for i in range(n - period, n):
            tr = max(
                self.high_history[i] - self.low_history[i],
                abs(self.high_history[i] - self.close_history[i - 1]),
                abs(self.low_history[i] - self.close_history[i - 1]),
            )
            trs.append(tr)
        self.intraday_atr = sum(trs) / period

    # ==================== Entry ====================

    def _enter_trade(self, close: float, side: OrderSide, direction: int,
                     stop: float, tp: float):
        num_contracts = self._compute_contracts()
        qty = Quantity.from_int(num_contracts)

        order = self.order_factory.market(
            instrument_id=self.instrument_id,
            order_side=side,
            quantity=qty,
            time_in_force=TimeInForce.GTC,
        )
        self.submit_order(order)

        self.entry_price = close
        self.stop_price = stop
        self.tp_price = tp
        self.trade_side = direction
        self.trades_today += 1

        label = "LONG" if direction == 1 else "SHORT"
        risk = abs(close - stop)
        self.log.info(
            f"{label}: entry={close:.2f}, stop={stop:.2f}, TP={tp:.2f}, "
            f"risk={risk:.2f}, fastEMA={self.fast_ema:.2f}, "
            f"slowEMA={self.slow_ema:.2f}, iATR={self.intraday_atr:.2f}"
        )

    # ==================== Position Management ====================

    def _manage_position(self, position: Position, high: float, low: float,
                         close: float):
        if self.trade_side == 1:  # LONG
            if low <= self.stop_price:
                self.close_position(position)
                self.log.info(f"STOP hit at {self.stop_price:.2f}")
                self._reset_trade_state()
                return
            if high >= self.tp_price:
                self.close_position(position)
                self.log.info(f"TP hit at {self.tp_price:.2f}")
                self._reset_trade_state()
                return
        else:  # SHORT
            if high >= self.stop_price:
                self.close_position(position)
                self.log.info(f"STOP hit at {self.stop_price:.2f}")
                self._reset_trade_state()
                return
            if low <= self.tp_price:
                self.close_position(position)
                self.log.info(f"TP hit at {self.tp_price:.2f}")
                self._reset_trade_state()
                return

    # ==================== Helpers ====================

    def _compute_contracts(self) -> int:
        cfg = self.config
        if cfg.dollars_per_contract <= 0:
            return cfg.contracts
        account = self.portfolio.account(self.instrument_id.venue)
        if account is None:
            return cfg.contracts
        equity = float(account.balance_total().as_double())
        return max(1, int(equity // cfg.dollars_per_contract))

    def _get_open_position(self):
        positions = self.cache.positions(instrument_id=self.instrument_id)
        if positions:
            for pos in positions:
                if not pos.is_closed:
                    return pos
        return None

    def _reset_trade_state(self):
        self.entry_price = 0.0
        self.stop_price = 0.0
        self.tp_price = 0.0
        self.trade_side = 0
